README.md
The Yammer Ruby Gem
====================
A Ruby wrapper for the Yammer REST APIs

This was originally forked from http://github.com/roadly/yammer.  I had to rename it due to namespace collisions with the latest Yam gem(which really users Yammer modules)

Installation
------------
    gem install yapper

What's new in 0.1.7?
----------------

Renames Yammer gem to Yapper
Updates the dependency on Yam gem to latest: 2.1.1

See HISTORY.md for more


Getting Started
---------------

You should probably be using Omniauth(http://github.com/intridea/omniauth), cuz its awesome, and be following that flow.  Ryan Bate's excellent RailsCasts are also a great reference: http://railscasts.com/episodes/236-omniauth-part-2

Whatever controller handles the callback, you will have access to a "credentials" key in the oauth object which has an "expires" key, and a "token" key.  I stash that entire object in a serialized string in my Authentication model.  This token is what you will need.

    token = current_user.authentications.find_by_provider("yammer").credentials.token
    yam = Yammer.new(oauth_token: token)
    
That's it!

    yam.groups
    yam.users


TODO
----

* Specs, specs and specs ... we lack tests even basic ones.
* Support for the following APIs
  * Likes
  * Topics
  * Users
  * Group Memberships
  * Networks
  
Contributing
------------
In the spirit of [free software](http://www.fsf.org/licensing/essays/free-sw.html), **everyone** is encouraged to help improve this project.

Here are some ways *you* can contribute:

* by using alpha, beta, and prerelease versions
* by reporting bugs
* by suggesting new features
* by writing or editing documentation
* by writing tests
* by writing code (**no patch is too small**: fix typos, add comments, clean up inconsistent whitespace)
* by refactoring code
* by closing [issues](https://github.com/roadly/yammer/issues)
* by reviewing patches

All contributors will be added to the [HISTORY](https://github.com/roadly/yammer/blob/master/HISTORY.md)
file and will receive the respect and gratitude of the community.

Submitting an Issue
-------------------
We use the [GitHub issue tracker](https://github.com/roadly/yammer/issues) to track bugs and
features. Before submitting a bug report or feature request, check to make sure it hasn't already
been submitted. You can indicate support for an existing issue by voting it up. When submitting a
bug report, please include a [Gist](https://gist.github.com/) that includes a stack trace and any
details that may be necessary to reproduce the bug, including your gem version, Ruby version, and
operating system. 

Submitting a Pull Request
-------------------------
1. Fork the project.
2. Create a topic branch.
3. Implement your feature or bug fix.
4. Add documentation for your feature or bug fix.
5. Run <tt>bundle exec rake doc:yard</tt>. If your changes are not 100% documented, go back to step 4.
6. Commit and push your changes.
7. Submit a pull request. Please do not include changes to the gemspec, version, or history file. (If you want to create your own version for some reason, please do so in a separate commit.)



Copyright
---------
Copyright (c) 2011, Peter Philips <pete <-at-> p373.net>
See [LICENSE](https://github.com/roadly/yammer/blob/master/LICENSE.md) for details.


HISTORY.md
0.1.7 - December 26th, 2013
----------------

* [synth](https://github.com/synth) Renames Yammer gem to Yapper
* [synth](https://github.com/synth) Updates the dependency on Yam gem to latest: 2.1.1

0.1.4
----------------

This release rips out the old api dependencies on Faraday and whatever else handles connection related issues and passed it all off to Yammer's own Yam gem(which itself uses Faraday).

Improved api support

    yam.users
    yam.in_group  

Better Exception Handling

    It was completely broken, now you can catch all sorts of exceptions with Yammer api based upon http status:
    
    - 400: Yammer::BadRequest
    - 401: Yammer::Unauthorized
    - 403: Yammer::Forbidden
    - 404: Yammer::NotFound
    - 406: Yammer::NotAcceptable
    - 500: Yammer:InternalServerError
    - 502: Yammer::BadGateway
    - 503: Yammer::ServiceUnavailable
    
0.1.2 - April 19th, 2012
-------------------------
* [ybenjo](https://github.com/ybenjo) added search, groups and suggestions api support.

0.1.1 - Sept 21st, 2011
-------------------------
* [Rob Warner](https://github.com/hoop33) added about_topic and user[:id]

0.1.0 - July 15th, 2011
-------------------------
* First public release
* Basic README.md

0.0.2 - July 9th, 2011
-------------------------
* Adding support for messages in threads

0.0.1 - July 9th, 2011
-------------------------
* [Initial release](http://github.com/roadly/yammer)

lib/yapper/error.rb
module Yapper

  # This gem seems to monkey patch faraday to catch certain exceptions
  # and even that attempt does not seem to be loaded anywhere, so instead
  # we'll catch the error on every request(and possibly elsewhere)
  # and then pass the exception to this class who will figure out the proper 
  # high level exception to raise
  class ErrorProxy
    def self.new(exception)
      return self.determine_exception(exception)
    end
    
    def self.determine_exception(exception)
      exception = exception
      message = exception.message
      status = exception.response[:status]
      headers = exception.response[:headers]

      case status.to_i
      when 400
        Yapper::BadRequest.new(error_message(message), headers)
      when 401
        Yapper::Unauthorized.new(error_message(message), headers)
      when 403
        Yapper::Forbidden.new(error_message(message), headers)
      when 404
        Yapper::NotFound.new(error_message(message), headers)
      when 429
        Yapper::TooManyRequests.new(error_message(message), headers)
      when 406
        Yapper::NotAcceptable.new(error_message(message), headers)
      when 500
        Yapper::InternalServerError.new(error_message("Something is technically wrong."), headers)
      when 502
        Yapper::BadGateway.new(error_message("Yammer is down or being upgraded."), headers)
      when 503
        Yapper::ServiceUnavailable.new(error_message("(__-){ Yammer is over capacity."), headers)
      else
        Exception.new("Unhandled Exception Status: #{status} - #{message}")
      end      
    end
  
    private
    def self.error_message(msg)
      #TODO: this can be used in the future to print nicer messages
      msg
    end
  end

  # Custom error class for rescuing from all Yammer errors
  class Error < StandardError
    attr_reader :http_headers

    def initialize(message, http_headers)
      @http_headers = Hash[http_headers]
      super message
    end

    def ratelimit_reset
      Time.at(@http_headers.values_at('x-ratelimit-reset', 'X-RateLimit-Reset').detect{|value| value}.to_i)
    end

    def ratelimit_limit
      @http_headers.values_at('x-ratelimit-limit', 'X-RateLimit-Limit').detect{|value| value}.to_i
    end

    def ratelimit_remaining
      @http_headers.values_at('x-ratelimit-remaining', 'X-RateLimit-Remaining').detect{|value| value}.to_i
    end

    def retry_after
      [(ratelimit_reset - Time.now).ceil, 0].max
    end
  end

  # Raised when Yammer returns the HTTP status code 400
  class BadRequest < Error; end

  # Raised when Yammer returns the HTTP status code 401
  class Unauthorized < Error; end

  # Raised when Yammer returns the HTTP status code 403
  class Forbidden < Error; end

  # Raised when Yammer returns the HTTP status code 404
  class NotFound < Error; end

  # Raised when Yammer returns the HTTP status code 406
  class NotAcceptable < Error; end

  # Raised when Yammer returns the HTTP status code 500
  class InternalServerError < Error; end

  # Raised when Yammer returns the HTTP status code 502
  class BadGateway < Error; end

  # Raised when Yammer returns the HTTP status code 503
  class ServiceUnavailable < Error; end

  # Raised when you hit Yammer's request rate limit
  class TooManyRequests < Error;end
end
